### PR TITLE
[NCL-2517] Use both external and internal URL in the import process

### PR DESCRIPTION
**Pull request by stanlyDoge**

### CODE REVIEW

File: bc-backend/src/main/java/org/jboss/da/bc/backend/impl/FinalizerImpl.java [hunk 1]
```diff
@@ -92,7 +92,13 @@ public Integer createBCs(int id, ProjectHiearchy toplevelBc) throws Communicatio
...
             ProjectDetail project = hiearchy.getProject();
             if (project.isUseExistingBc()) {
-                List<BuildConfiguration> existingBcs = bcFinder.lookupBcByScm(project.getScmUrl(), project.getScmRevision());
+                List<BuildConfiguration> existingBcs;
+                if (project.getInternalScmUrl() == null) {
+                    existingBcs = bcFinder.lookupBcByScm(project.getExternalScmUrl(), project.getExternalScmRevision());
+                }
+                else {
+                    existingBcs = bcFinder.lookupBcByScm(project.getInternalScmUrl(), project.getInternalScmRevision());
+                }
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/backend/impl/FinalizerImpl.java]: create bcFinder.lookupBcByInternalSCM and bcFinder.lookupBcByExternalSCM and combine the lists

File: bc-backend/src/main/java/org/jboss/da/bc/impl/BuildConfigurationGeneratorImpl.java [hunk 2]
```diff
@@ -161,9 +161,11 @@ private boolean validate(ProjectHiearchy hiearchy) throws IllegalStateException
                 noerror = false;
             }
 
-            if (project.getScmUrl() == null || project.getScmUrl().isEmpty()) {
-                project.addError(BcError.NO_SCM_URL);
-                noerror = false;
+            if (project.getInternalScmUrl() == null || project.getInternalScmUrl().isEmpty()) {
+                    if (project.getExternalScmUrl() == null || project.getExternalScmUrl().isEmpty()) {
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/impl/BuildConfigurationGeneratorImpl.java]: Indentation !!!!!!

File: bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java [hunk 3]
```diff
@@ -123,9 +130,15 @@ private void setDependencies(ProjectHiearchy hiearchy) {
         } catch (FindGAVDependencyException ex) {
             ProjectDetail project = toplevel.getProject();
             try {
-                // try to get dependencies from scm url instead
-                dependencies = depGenerator.getToplevelDependencies(project.getScmUrl(),
-                        project.getScmRevision(), gav);
+                if (project.getInternalScmUrl() == null) {
+                    // try to get dependencies from scm url instead
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java]: maybe enhance the comment with statement that we prefer internal first

File: bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java [hunk 4]
```diff
@@ -209,20 +227,30 @@ private void setSCMInfo(ProjectDetail project, Optional<POMInfo> pomInfo) {
 
         if (!url.isPresent() || !rev.isPresent()) {
             try {
-                boolean gavInRepository = scm.isGAVInRepository(scmUrl, revison, project.getGav());
+                String ScmUrl;
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java]: variable name should start with lowercase

**Comment by janinko** [comment]
retest this please

**Comment by stanlyDoge** [comment]
retest this please

**Comment by stanlyDoge** [comment]
retest this please

**Comment by janinko** [comment]
retest this please

**Comment by stanlyDoge** [comment]
retest this please

**Comment by janinko** [comment]
retest this please

File: communication/src/main/java/org/jboss/da/communication/pnc/impl/PNCConnectorImpl.java [hunk 5]
```diff
@@ -243,7 +243,7 @@ public BuildConfigurationSet createBuildConfigurationSet(BuildConfigurationSet b
                 // TODO solve pagination
                 "build-configurations?q=scmRepoURL=='%s';scmRevision=='%s'&pageIndex=0&pageSize=500",
                         scmUrl, scmRevision);
-
+        System.out.println("URL dddddddddd " + requestUrl);
```
Comment (janinko) [inline communication/src/main/java/org/jboss/da/communication/pnc/impl/PNCConnectorImpl.java]: wut?

File: bc-rest/src/main/java/org/jboss/da/bc/facade/AbstractBuildConfigurationsFacade.java [hunk 6]
```diff
@@ -140,8 +144,15 @@ protected void fillInfoEntity(InfoEntity ie, GeneratorEntity ge) {
...
-        String revision = ie.getTopLevelBc().getScmRevision();
+        String url = "";
+        String revision = "";
+        if (ie.getTopLevelBc().getScmUrl() == null) {
+            url = ie.getTopLevelBc().getExternalScmUrl();
+            revision = ie.getTopLevelBc().getExternalScmRevision();
+        } else {
+            url = ie.getTopLevelBc().getScmUrl();
+            revision = ie.getTopLevelBc().getScmRevision();
+        }
```
Comment (janinko) [inline bc-rest/src/main/java/org/jboss/da/bc/facade/AbstractBuildConfigurationsFacade.java]: This will always set ``GeneratorEntity.toplevelBc.scmUrl`` and ``GeneratorEntity.toplevelBc.scmUrl`` (ie the internal scm coordinates), regadless of the fact that ``InfoEntity.topLevelBc`` could have internal or external coordinates.

Comment (janinko) [inline bc-rest/src/main/java/org/jboss/da/bc/facade/AbstractBuildConfigurationsFacade.java]: There will have to be some changes in GeneratorEntity too.

File: bc-backend/src/main/java/org/jboss/da/bc/model/backend/ProjectDetail.java [hunk 7]
```diff
@@ -24,11 +24,19 @@
 
     @Getter
     @Setter
-    private String scmUrl; // required
+    protected String scmUrl;
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/model/backend/ProjectDetail.java]: Why the change of visibility?

File: bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java [hunk 8]
```diff
@@ -209,20 +226,30 @@ private void setSCMInfo(ProjectDetail project, Optional<POMInfo> pomInfo) {
...
+                                url = Optional.of(scmUrlTemp);
                             }
                         } else {
-                            url = Optional.of(scmUrl);
-                            rev = Optional.of(revison);
+                            url = Optional.of(scmUrlTemp);
+                            rev = Optional.of(revisionTemp);
                         }
                     }
                 }
```
Comment (janinko) [inline bc-backend/src/main/java/org/jboss/da/bc/impl/ProjectHiearchyCreator.java]: You should try and consider where the scm url was taken from (external internal) and at the end set it properly.
 * If the gav is in repository
   * we should copy the coordinates (internal and external) from the toplevel
 * if the gav is not in repository but there is coordinates in the pom
   * we should consider the url from pom to be external url
 * if the gav is not in repository and there are no coordinates in the pom
  * both internal and external url are null

File: bc-backend/src/main/java/org/jboss/da/bc/model/backend/ProjectDetail.java (hunk 7 above)
Comment (stanlyDoge) [inline bc-backend/src/main/java/org/jboss/da/bc/model/backend/ProjectDetail.java]: pr and Ctrl + space i guess :grimacing:

File: bc-rest/src/main/java/org/jboss/da/bc/facade/BuildConfigurationsProductFacade.java [hunk 9]
```diff
@@ -30,6 +30,11 @@
     @Override
     protected ProductInfoEntity start(SCMLocator scm, EntryEntity entry) throws ScmException,
             PomAnalysisException, CommunicationException {
+        Optional<String> scmUrl = Optional.of(entry.getScmUrl());
+        scm.setInternal(scmUrl.isPresent());
+        /*
+         * if (entry.getScmUrl() == null) { scm.setInternal(false); } else { scm.setInternal(true); }
+         */
```
Comment (janinko) [inline bc-rest/src/main/java/org/jboss/da/bc/facade/BuildConfigurationsProductFacade.java]: Commented code

File: communication/src/main/java/org/jboss/da/communication/pnc/impl/PNCConnectorImpl.java [hunk 10]
```diff
@@ -243,7 +243,6 @@ public BuildConfigurationSet createBuildConfigurationSet(BuildConfigurationSet b
                 // TODO solve pagination
                 "build-configurations?q=scmRepoURL=='%s';scmRevision=='%s'&pageIndex=0&pageSize=500",
                         scmUrl, scmRevision);
-
         ClientResponse<PNCResponseWrapper<List<BuildConfiguration>>> response = get(requestUrl,
```
Comment (janinko) [inline communication/src/main/java/org/jboss/da/communication/pnc/impl/PNCConnectorImpl.java]: whitespace change in otherwise unchanged file
